src/internal/partial_solution.rs

```diff
@@ -94,19 +100,39 @@ impl<P: Package, V: Version> PartialSolution<P, V> {
     /// This should be a package with a positive derivation but no decision yet.
     /// If multiple choices are possible, use a heuristic.
     ///
-    /// Pub chooses the package with the fewest versions
+    /// We and Pub chooses the package with the fewest versions
     /// matching the outstanding constraint.
     /// This tends to find conflicts earlier if any exist,
     /// since these packages will run out of versions to try more quickly.
     /// But there's likely room for improvement in these heuristics.
     ///
-    /// Here we just pick the first one.
     /// TODO: improve? (do not introduce any side effect if trying to improve)
-    pub fn pick_package(&self) -> Option<(P, Term<V>)> {
-        self.memory
+    pub fn pick_package(
+        &self,
+        dependency_provider: &impl DependencyProvider<P, V>,
+    ) -> Result<Option<(P, Term<V>)>, PubGrubError<P, V>> {
+        let mut out: Option<(P, Term<V>)> = None;
+        let mut min_key = usize::MAX;
+        for (p, term) in self
+            .memory
             .potential_packages()
-            .next()
-            .map(|(p, all_terms)| (p.clone(), Term::intersect_all(all_terms.iter())))
+            .map(|(p, all_terms)| (p, Term::intersect_all(all_terms.iter())))
```

Comment:
This looks like a very pricey thing to do if we effectively iterate over all potential packages. In such case, it would probably be worth keeping in `memory` not a `Vec` of derivations but their intersections directly. I haven't reread the code but I think we may never need individual derivations for anything. Every usage we have of `Memory` could be as efficient or more if we actually had the intersection directly.

To be checked

Reply:
Yes, That is definitely worth looking into! But with this already being 30x it is hard to tell based on this benchmark. It seams to worth committing now and evaluating that again when we have a new benchmark.

Reply:
Alright!